python/cog/server/clients.py

```diff
@@ -0,0 +1,319 @@
+# is tuned below 100ms. This should help us get output sooner for models that
+# are latency sensitive.
+SKIP_START_EVENT = _response_interval < 0.1
+
+WebhookSenderType = Callable[[Any, WebhookEvent], Awaitable[None]]
+
+
+def common_headers() -> "dict[str, str]":
+    headers = {"user-agent": _user_agent}
+    return headers
+
+
+def webhook_headers() -> "dict[str, str]":
+    headers = common_headers()
+    auth_token = os.environ.get("WEBHOOK_AUTH_TOKEN")
+    if auth_token:
+        headers["authorization"] = "Bearer " + auth_token
+
+    return headers
+
+
+async def on_request_trace_context_hook(request: httpx.Request) -> None:
+    ctx = current_trace_context() or {}
+    request.headers.update(cast(Mapping[str, str], ctx))
+
+
+def httpx_webhook_client() -> httpx.AsyncClient:
+    return httpx.AsyncClient(headers=webhook_headers(), follow_redirects=True)
+
+
+def httpx_retry_client() -> httpx.AsyncClient:
+    # This session will retry requests up to 12 times, with exponential
+    # backoff. In total it'll try for up to roughly 320 seconds, providing
+    # resilience through temporary networking and availability issues.
+    transport = RetryTransport(
+        max_attempts=12,
+        backoff_factor=0.1,
+        retry_status_codes=[429, 500, 502, 503, 504],
+        retryable_methods=["POST"],
+    )
+    return httpx.AsyncClient(
+        event_hooks={"request": [on_request_trace_context_hook]},
+        headers=webhook_headers(),
+        transport=transport,
+        follow_redirects=True,
+    )
+
+
+def httpx_file_client() -> httpx.AsyncClient:
+    # verify: Union[str, bool, ssl.SSLContext] = True
+    transport = RetryTransport(
+        max_attempts=3,
+        backoff_factor=0.1,
+        retry_status_codes=[408, 429, 500, 502, 503, 504],
+        retryable_methods=["PUT"],
+        verify=os.environ.get("CURL_CA_BUNDLE", True),
+    )
+    # set connect timeout to slightly more than a multiple of 3 to avoid
+    # aligning perfectly with TCP retransmission timer
+    # requests has no write timeout, keep that
+    # httpx default for pool is 5, use that
+    timeout = httpx.Timeout(connect=10, read=15, write=None, pool=5)
+
+    return httpx.AsyncClient(
+        event_hooks={"request": [on_request_trace_context_hook]},
+        headers=common_headers(),
+        transport=transport,
+        follow_redirects=True,
+        timeout=timeout,
+        http2=True,
+    )
+
+
+class ChunkFileReader:
+    def __init__(self, fh: io.IOBase) -> None:
+        self.fh = fh
+
+    async def __aiter__(self) -> AsyncIterator[bytes]:
+        self.fh.seek(0)
+        while True:
+            chunk = self.fh.read(1024 * 1024)
+            if isinstance(chunk, str):
+                chunk = chunk.encode("utf-8")
+            if not chunk:
+                log.info("finished reading file")
+                break
+            yield chunk
+
+
+# there's a case for splitting this apart or inlining parts of it
```

Comment:
I don't think HTTP clients and data structure manipulations should be smooshed together like this. We should extract these back into separate business logic and client management pieces.

The current approach makes it hard to test, hard to iterate on (e.g. `TODO: upload concurrently` would be far easier to achieve if we separated the processes of working out what we have to upload from the process of actually uploading it), and hard to understand.

Reply:
I'm not entirely convinced, this is identical to json.upload_files. I am sympathetic to extracting some helper functions for data structure manipulations. 

you could have a helper function yield each file that has to be uploaded, but then those have to go back into the data structure anyway, it doesn't seem like less complexity to me.

the previous setup had about five layers of indirection across four files

Reply:
that said if you have a narrow proposal I would be open to that